### ISSUE TITLE
[Bug] Cannot freeze an expression imported with custom headers

### ISSUE THREAD

**Issue by ari-becker** [dhall-lang/dhall-haskell#1093]
Version: 1.24.0

Steps to reproduce:
1. create `broken.dhall`:
```dhall
{ Prelude =
    https://prelude.dhall-lang.org/package.dhall using [ { header =
                                                             "Hello"
                                                         , value =
                                                             "World"
                                                         }
                                                       ]
}
```
2. Verify that `broken.dhall` parses correctly by calling `dhall --file broken.dhall`
3. call `dhall freeze --inplace broken.dhall`. Observe that the call returns successfully. The contents of `broken.dhall`, as expected, are now:
```dhall
{ Prelude =
    https://prelude.dhall-lang.org/package.dhall using [ { header =
                                                             "Hello"
                                                         , value =
                                                             "World"
                                                         }
                                                       ] sha256:0c04cbe34f1f2d408e8c8b8cb0aa3ff4d5656336910f7e86190a6d14326f966d
}
```
4. call `dhall --file broken.dhall`, and receive the following error:
```text

↳ https://prelude.dhall-lang.org/package.dhall using [ { header = "Hello", value = "World" } ] sha256 : 0 c04cbe34f1f2d408e8c8b8cb0aa3ff4d5656336910f7e86190a6d14326f966d

Error: Not a function

7:                                                                 0c04cbe34f1f2d408e8c8b8cb0aa3ff4d5656336910f7e86190a6d14326f966d

(stdin):7:65

2:     https://prelude.dhall-lang.org/package.dhall using [ { header =
3:                                                              "Hello"
4:                                                          , value =
5:                                                              "World"
6:                                                          }
7:                                                        ] sha256:0c04cbe34f1f2d408e8c8b8cb0aa3ff4d5656336910f7e86190a6d14326f966d

(stdin):2:5

```

Not sure if this is a bug in the Haskell implementation or the language specification.

**Comment by EggBaconAndSpam** [comment]
That seems to be a bug in the 1.24.0 parser. The bug doesn't occur with the latest git version.

**Comment by Gabriel439** [comment]
Yeah, this was a bug in the parser.  This was fixed in 04d82120a415569879bf7faf044c432e76ab6f0e.

The issue was that before that change the parser was not following the grammar for the `using` clause of an import because it permitted an arbitrary expression (i.e. `completeExpression`) instead of an `importExpression`.  The former permits unparenthesized type annotations whereas the latter does not.

Because the former implementation permitted unparenthesized type annotations the `sha256:0c04cbe34f1f2d408e8c8b8cb0aa3ff4d5656336910f7e86190a6d14326f966d` was parsed as `sha256 : 0 c04cbe34f1f2d408e8c8b8cb0aa3ff4d5656336910f7e86190a6d14326f966d` (i.e. the variable named `sha256` annotated with a type of `0 c04cbe34f1f2d408e8c8b8cb0aa3ff4d5656336910f7e86190a6d14326f966d`), which led to a type-checking failure.

I'll mark this closed since it's now fixed in `master`, which now matches the official grammar in this regard.
